File: go/stripe/stripe.go
package stripe

import (
	"context"
	"strings"

	cekat "golang.cekat.ai/event-sdk"
)

const visitorMetadataKey = "cekat_visitor_id"

func MetadataForVisitor(visitorID string) map[string]string {
	visitorID, ok := validVisitorID(visitorID)
	if !ok {
		return map[string]string{}
	}
	return map[string]string{visitorMetadataKey: visitorID}
}

func MetadataFromContext(ctx context.Context) map[string]string {
	visitorID, ok := cekat.VisitorIDFromContext(ctx)
	if !ok {
		return map[string]string{}
	}
	return MetadataForVisitor(visitorID)
}

func MergeMetadata(metadata map[string]string, visitorID string) map[string]string {
	merged := make(map[string]string, len(metadata)+1)
	for key, value := range metadata {
		merged[key] = value
	}
	if normalizedVisitorID, ok := validVisitorID(visitorID); ok {
		merged[visitorMetadataKey] = normalizedVisitorID
	}
	return merged
}

func validVisitorID(value string) (string, bool) {
	value = strings.TrimSpace(value)
	if len(value) == 0 || len(value) > 128 {
		return "", false
	}
	for _, character := range value {
		if (character < 'a' || character > 'z') && (character < 'A' || character > 'Z') && (character < '0' || character > '9') && character != '_' && character != '-' {
			return "", false
		}
	}
	return value, true
}

File: node/src/node/stripe.ts
import { currentVisitorId } from './visitor-context.js';

const visitorMetadataKey = 'cekat_visitor_id';

export function metadataForVisitor(visitorId: string | undefined): Record<string, string> {
  const normalizedVisitorId = validVisitorId(visitorId);
  return normalizedVisitorId === undefined ? {} : { [visitorMetadataKey]: normalizedVisitorId };
}

export function metadataFromCurrentVisitor(): Record<string, string> {
  return metadataForVisitor(currentVisitorId());
}

export function mergeMetadata(metadata: Readonly<Record<string, string>>, visitorId: string | undefined): Record<string, string> {
  const merged = { ...metadata };
  const normalizedVisitorId = validVisitorId(visitorId);
  if (normalizedVisitorId !== undefined) merged[visitorMetadataKey] = normalizedVisitorId;
  return merged;
}

function validVisitorId(visitorId: string | undefined): string | undefined {
  if (visitorId === undefined) return undefined;
  const normalizedVisitorId = visitorId.trim();
  return /^[A-Za-z0-9_-]{1,128}$/.test(normalizedVisitorId) ? normalizedVisitorId : undefined;
}
